fl4health/clients/apfl_client.py

```diff
@@ -1,255 +1,149 @@
-
-    def evaluate(self, parameters: NDArrays, config: Config) -> Tuple[float, int, Dict[str, Scalar]]:
-        if not self.initialized:
-            self.setup_client(config)
-
-        self.set_parameters(parameters, config)
-        # Default APFL uses an average meter
-        global_meter = MetricAverageMeter(self.metrics, "global")
-        local_meter = MetricAverageMeter(self.metrics, "local")
-        personal_meter = MetricAverageMeter(self.metrics, "personal")
-        loss, metric_values = self.validate(global_meter, local_meter, personal_meter)
-        # EvaluateRes should return the loss, number of examples on client, and a dictionary holding metrics
-        # calculation results.
-        return (
-            loss,
-            self.num_examples["validation_set"],
-            metric_values,
-        )
-
-    def _handle_logging(
-        self, loss_dict: Dict[str, float], metrics_dict: Dict[str, Scalar], is_validation: bool = False
-    ) -> None:
-        loss_string = "\t".join([f"{key}: {str(val)}" for key, val in loss_dict.items()])
-        metric_string = "\t".join([f"{key}: {str(val)}" for key, val in metrics_dict.items()])
-        metric_prefix = "Validation" if is_validation else "Training"
-        log(
-            INFO,
-            f"alpha: {self.model.alpha} \n"
-            f"Client {metric_prefix} Losses: {loss_string} \n"
-            f"Client {metric_prefix} Metrics: {metric_string}",
-        )
-
-    def train_step(self, input: torch.Tensor, target: torch.Tensor) -> ApflTrainStepOutputs:
+
+        # Need to track total_steps across rounds for WANDB reporting
+        self.total_steps: int = 0
+
+    def is_start_of_local_training(self, step: int) -> bool:
+        return step == 0
+
+    def update_after_step(self, step: int) -> None:
+        """
+        Called after local train step on client. step is an integer that represents
+        the local training step that was most recently completed.
+        """
+        if self.is_start_of_local_training(step) and self.model.adaptive_alpha:
+            self.model.update_alpha()
+
+    def setup_client(self, config: Config) -> None:
+        """
+        Set dataloaders, optimizers, parameter exchangers and other attributes derived from these.
+        """
+        model = self.get_model(config)
+        assert isinstance(model, ApflModule)
+        self.model = model
+        train_loader, val_loader = self.get_data_loaders(config)
+        self.train_loader = train_loader
+        self.val_loader = val_loader
+
+        # The following lines are type ignored because torch datasets are not "Sized"
+        # IE __len__ is considered optionally defined. In practice, it is almost always defined
+        # and as such, we will make that assumption.
+        self.num_train_samples = len(self.train_loader.dataset)  # type: ignore
+        self.num_val_samples = len(self.val_loader.dataset)  # type: ignore
+
+        optimizer_dict = self.get_optimizer(config)
```

Comment:
One way that we could get around having to redefine the setup_client for this is to have a `set_optimizers(config)` function which by default does what the BasicClient does, but in APFL is overridden to perform this operation instead. Then setup client could remain un-duplicated (the `assert isinstance(model, ApflModule)` could just be called before the `super().setup_client(config)` call

Reply:
Good call! Updated in [9959c89](https://github.com/VectorInstitute/FL4Health/pull/62/commits/9959c896eba8b13846f86a313be23b34619dc1cb)